### ISSUE TITLE
RPMB support currently broken on 3.18.0 with imx8mm-evk (fused/closed)

### ISSUE THREAD

**Issue by ricardosalveti** [OP-TEE/optee_os#5501]
I'm updating my recipes to use the latest OP-TEE release, but I'm facing an interesting issue on imx8mm-evk (fused/closed) which is causing RPMB to not be accessible (no matching device key).

At first I thought it would be HUK related, but when comparing with 3.17 I'm getting the same HUK value. Then I started bisecting 
and found that it stopped working at https://github.com/OP-TEE/optee_os/commit/7c3ab7744d38842cfc4fd2cc095b6a489a2c28bb, which is surprising.

Adding a bit of debug the issue seems to happen on huk_subkey_derive, which is returning a zeroed subkey, instead of a valid one (huk and const data are both the same).

When testing commit https://github.com/OP-TEE/optee_os/commit/5cc089856c2fa78523320e3b31dea48c2d42b920:
```
D/TC:? 0 huk_subkey_derive:97 const data
D/TC:? 0 huk_subkey_derive:98 00000000be10eb18  45 01 00 44 47 34 30 31  36 00 8d ce 54 0b 26 00
D/TC:? 0 huk_subkey_derive:104 subkey
D/TC:? 0 huk_subkey_derive:105 00000000be0ecd90  2d 3a 5e cc d1 b5 8d 29  fb bc ff 04 7e 75 7e b6
D/TC:? 0 huk_subkey_derive:105 00000000be0ecda0  75 7a 97 f4 7a 78 89 43  95 bd 43 29 35 37 90 00
```

Now with https://github.com/OP-TEE/optee_os/commit/7c3ab7744d38842cfc4fd2cc095b6a489a2c28bb:
```
D/TC:? 0 huk_subkey_derive:97 const data
D/TC:? 0 huk_subkey_derive:98 00000000be10eb18  45 01 00 44 47 34 30 31  36 00 8d ce 54 0b 26 00
D/TC:? 0 huk_subkey_derive:104 subkey
D/TC:? 0 huk_subkey_derive:105 00000000be0ecd90  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
D/TC:? 0 huk_subkey_derive:105 00000000be0ecda0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
```

Basically ``crypto_mac_final`` seems to be broken in a way (without returning any error).

Another interesting data is that the diff bellow also fixes the behavior (partially reverting https://github.com/OP-TEE/optee_os/commit/7c3ab7744d38842cfc4fd2cc095b6a489a2c28bb):
```
diff --git a/core/mm/core_mmu.c b/core/mm/core_mmu.c
index 6fae580cb..cd73edf1c 100644
--- a/core/mm/core_mmu.c
+++ b/core/mm/core_mmu.c
@@ -696,7 +696,7 @@ static void add_va_space(struct tee_mmap_region *memory_map, size_t num_elems,
 uint32_t core_mmu_type_to_attr(enum teecore_memtypes t)
 {
        const uint32_t attr = TEE_MATTR_VALID_BLOCK;
-       const uint32_t tagged = TEE_MATTR_MEM_TYPE_TAGGED <<
+       const uint32_t tagged = TEE_MATTR_MEM_TYPE_CACHED <<
                                TEE_MATTR_MEM_TYPE_SHIFT;
        const uint32_t cached = TEE_MATTR_MEM_TYPE_CACHED <<
                                TEE_MATTR_MEM_TYPE_SHIFT;
```

**Comment by ricardosalveti** [comment]
@jenswi-linaro maybe you have an idea here.

**Comment by ricardosalveti** [comment]
The commit says "The mappings used by TEE_MATTR_MEM_TYPE_TAGGED is replaced by TEE_MATTR_MEM_TYPE_CACHED if MTE isn't supported or configured" but I don't see how that is currently being done, as it just changed from cached to tagged directly.

**Comment by jenswi-linaro** [comment]
> The commit says "The mappings used by TEE_MATTR_MEM_TYPE_TAGGED is replaced by TEE_MATTR_MEM_TYPE_CACHED if MTE isn't supported or configured" but I don't see how that is currently being done, as it just changed from cached to tagged directly.

Looking at the arm64 implementation of `core_init_mmu_regs()` you can see how the value for mair_el1 is initialized. A new index is defined for tagged memory, but all settings used are still normal cached memory.

It's a good thing that you managed to bisect to this commit since it's not supposed to change behavior except that another index in MAIR_EL1 is used for some mappings. From the error, it looks like some cache maintenance hasn't been done properly. It seems that something is confused by the new index.

**Comment by ricardosalveti** [comment]
> Looking at the arm64 implementation of `core_init_mmu_regs()` you can see how the value for mair_el1 is initialized. A new index is defined for tagged memory, but all settings used are still normal cached memory.

Indeed, the logic to enable is available in another commit.

> It's a good thing that you managed to bisect to this commit since it's not supposed to change behavior except that another index in MAIR_EL1 is used for some mappings. From the error, it looks like some cache maintenance hasn't been done properly. It seems that something is confused by the new index.

But funny that the issue seems to happen when setting up `desc`. If I just revert cached -> tagged (normal) for MEM_AREA_TEE_ASAN, it works fine:
```
diff --git a/core/mm/core_mmu.c b/core/mm/core_mmu.c
index 6fae580cb..4469a793a 100644
--- a/core/mm/core_mmu.c
+++ b/core/mm/core_mmu.c
@@ -717,7 +717,7 @@ uint32_t core_mmu_type_to_attr(enum teecore_memtypes t)
        case MEM_AREA_NEX_RAM_RO: /* This has to be r/w during init runtime */
        case MEM_AREA_NEX_RAM_RW:
        case MEM_AREA_TEE_ASAN:
-               return attr | TEE_MATTR_SECURE | TEE_MATTR_PRW | tagged;
+               return attr | TEE_MATTR_SECURE | TEE_MATTR_PRW | cached;
        case MEM_AREA_TEE_COHERENT:
                return attr | TEE_MATTR_SECURE | TEE_MATTR_PRWX | noncache;
        case MEM_AREA_TA_RAM:
```

**Comment by ricardosalveti** [comment]
And another interesting point is that CFG_CORE_SANITIZE_KADDRESS and CFG_WITH_PAGER are both disabled, so `register_phys_mem_ul` is not called for MEM_AREA_TEE_ASAN.

**Comment by ricardosalveti** [comment]
MEM_AREA_TEE_RAM_RW:
```
diff --git a/core/mm/core_mmu.c b/core/mm/core_mmu.c
index 6fae580cb..b98056e4e 100644
--- a/core/mm/core_mmu.c
+++ b/core/mm/core_mmu.c
@@ -714,6 +714,7 @@ uint32_t core_mmu_type_to_attr(enum teecore_memtypes t)
        case MEM_AREA_INIT_RAM_RO:
                return attr | TEE_MATTR_SECURE | TEE_MATTR_PR | tagged;
        case MEM_AREA_TEE_RAM_RW:
+               return attr | TEE_MATTR_SECURE | TEE_MATTR_PRW | cached;
        case MEM_AREA_NEX_RAM_RO: /* This has to be r/w during init runtime */
        case MEM_AREA_NEX_RAM_RW:
        case MEM_AREA_TEE_ASAN:
```

Gets it to work again.

**Comment by ricardosalveti** [comment]
And for curiosity, enabling `CFG_MEMTAG=y` causes OP-TEE to not even boot (imx8mm is A53), which also seems to be unexpected from the commit that adds support for MTE.

**Comment by ricardosalveti** [comment]
Let me know if there is anything in particular that you think that could be useful to debug this one.

**Comment by jenswi-linaro** [comment]
There has to be some mismatch with the MAIR index. Perhaps double-check that MAIR_EL1 indeed is what we expect. Do a write_at_s1e1r() for some interesting address and look at the returned memory properties.

**Comment by ricardosalveti** [comment]
The issue is fixed by 3077e8122acaff13a1b7ec49e95a5b54e444abda, so all good.
